File: backend/app/api/endpoints/wishlist.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List
from ...db.mongodb import get_database
from ...models.schemas import WishlistItem, WishlistItemCreate, User, Product
from ..deps import get_current_user

router = APIRouter()

@router.get("/", response_model=List[dict])
async def get_wishlist(current_user: User = Depends(get_current_user), db = Depends(get_database)):
    wishlist_items = await db.wishlist_items.find({"user_id": current_user.id}).to_list(1000)
    
    # Enrich with product details
    enriched_wishlist = []
    for item in wishlist_items:
        product = await db.products.find_one({"id": item["product_id"]})
        if product:
            enriched_wishlist.append({
                "wishlist_item": WishlistItem(**item),
                "product": Product(**product)
            })
    return enriched_wishlist

@router.post("/toggle", response_model=dict)
async def toggle_wishlist(
    item_in: WishlistItemCreate, 
    current_user: User = Depends(get_current_user), 
    db = Depends(get_database)
):
    # Check if product exists
    product = await db.products.find_one({"id": item_in.product_id})
    if not product:
        raise HTTPException(status_code=404, detail="Product not found")
    
    # Check if item already in wishlist
    existing_item = await db.wishlist_items.find_one({
        "user_id": current_user.id,
        "product_id": item_in.product_id
    })
    
    if existing_item:
        await db.wishlist_items.delete_one({"id": existing_item["id"]})
        return {"status": "removed", "message": "Product removed from wishlist"}
    else:
        wishlist_item = WishlistItem(
            user_id=current_user.id,
            product_id=item_in.product_id
        )
        await db.wishlist_items.insert_one(wishlist_item.dict())
        return {"status": "added", "message": "Product added to wishlist"}

@router.delete("/{item_id}")
async def remove_from_wishlist(
    item_id: str, 
    current_user: User = Depends(get_current_user), 
    db = Depends(get_database)
):
    result = await db.wishlist_items.delete_one({"id": item_id, "user_id": current_user.id})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Wishlist item not found")
    return {"message": "Item removed from wishlist"}

File: backend/src/routes/wishlist.js
const express = require('express');
const router = express.Router();
const adapter = require('../db/adapter');
const { authMiddleware } = require('../middleware/auth');

// Get wishlist (enriched with product details)
router.get('/', authMiddleware, async (req, res) => {
  const items = await adapter.getWishlistByUser(req.user.id);
  const enriched = [];
  for (const item of items) {
    const product = await adapter.findProductById(item.product_id);
    if (product) enriched.push({ wishlist_item: item, product });
  }
  res.json(enriched);
});

// Toggle wishlist item
router.post('/toggle', authMiddleware, async (req, res) => {
  const { product_id } = req.body;
  if (!product_id) return res.status(400).json({ detail: 'product_id required' });
  // check product exists
  const product = await adapter.findProductById(product_id);
  if (!product) return res.status(404).json({ detail: 'Product not found' });
  const r = await adapter.toggleWishlistItem(req.user.id, product_id);
  if (r.status === 'added') return res.json({ status: 'added', item: r.item });
  return res.json({ status: 'removed' });
});

// Remove from wishlist by id
router.delete('/:item_id', authMiddleware, async (req, res) => {
  const ok = await adapter.deleteWishlistItemById(req.params.item_id, req.user.id);
  if (!ok) return res.status(404).json({ detail: 'Wishlist item not found' });
  res.json({ message: 'Item removed from wishlist' });
});

module.exports = router;